### PR TITLE
Wib db id gen (PLEASE CREATE BRANCH wip-db-id-gen)

### PR DESCRIPTION
**Pull request by dartartem**

### CODE REVIEW

File: _build-and-test-all.sh [hunk 1]
```diff
@@ -32,4 +33,16 @@ echo 'show dbs' |  ./mongodb-cli.sh -i
 
 ./gradlew :end-to-end-tests:cleanTest :end-to-end-tests:test
 
+cat "add-database-id-support-to-eventuate-mysql.migration.sql" | ./mysql-cli.sh -i
```
Comment (cer) [inline _build-and-test-all.sh:36]: Shouldn't this only be executed when DATABASE == mysql?

File: dbid.env [hunk 2]
```diff
@@ -0,0 +1 @@
+EVENTUATE_OUTBOX_ID=1
```
Comment (cer) [inline dbid.env:1]: Please rename this file to `docker-compose-env-files/db-id-gen.env`. 
Ditto for `empty.env`

File: _build-and-test-all.sh (hunk 1 above)
Comment (dartartem) [reply _build-and-test-all.sh:36]: yes, I was thinking to add all conditions when we discuss mysql, but ok, I will do it now

File: dbid.env (hunk 2 above)
Comment (dartartem) [reply dbid.env:1]: ok

File: _build-and-test-all.sh [hunk 3]
```diff
@@ -32,4 +33,18 @@ echo 'show dbs' |  ./mongodb-cli.sh -i
...
 
+if [ "${DATABASE}" == "mysql" ]; then
+  cat "add-database-id-support-to-eventuate-mysql.migration.sql" | ./mysql-cli.sh -i
+fi
+
+manual_compose="docker-compose -f docker-compose-mysql-binlog.yml"
+services_to_restart="order-service customer-service order-history-service"
+
+${manual_compose} stop ${services_to_restart}
+${manual_compose} rm --force ${services_to_restart}
```
Comment (cer) [inline _build-and-test-all.sh:44]: Why is this necessary? Wont Docker Compose recreate the container anyway because it's been reconfigured with different env vars?

File: _build-and-test-all.sh [hunk 4]
```diff
@@ -32,4 +33,18 @@ echo 'show dbs' |  ./mongodb-cli.sh -i
...
 ./gradlew :end-to-end-tests:cleanTest :end-to-end-tests:test
 
+if [ "${DATABASE}" == "mysql" ]; then
+  cat "add-database-id-support-to-eventuate-mysql.migration.sql" | ./mysql-cli.sh -i
+fi
+
+manual_compose="docker-compose -f docker-compose-mysql-binlog.yml"
+services_to_restart="order-service customer-service order-history-service"
+
+${manual_compose} stop ${services_to_restart}
```
Comment (cer) [inline _build-and-test-all.sh:43]: Shouldn't the CDC be stopped too? It needs to be configured with the outbox ID

File: _build-and-test-all.sh (hunk 3 above)
Comment (dartartem) [reply _build-and-test-all.sh]: I did not think about it. But yes, I tested. PR is updated.

File: _build-and-test-all.sh (hunk 4 above)
Comment (dartartem) [reply _build-and-test-all.sh]: Current implementation requires to specify EVENTUATE_OUTBOX_ID immediately after upgrade even if it is not used.

File: _build-and-test-all.sh [hunk 5]
```diff
@@ -32,4 +33,12 @@ echo 'show dbs' |  ./mongodb-cli.sh -i
 
 ./gradlew :end-to-end-tests:cleanTest :end-to-end-tests:test
 
+if [ "${DATABASE}" == "mysql" ]; then
+  cat "add-database-id-support-to-eventuate-mysql.migration.sql" | ./mysql-cli.sh -i
+fi
+
+${dockerall}Up -P envFile=docker-compose-env-files/db-id-gen.env
```
Comment (cer) [inline _build-and-test-all.sh:40]: Why run the tests again if the database is not mysql?

File: docker-compose-mysql-binlog.yml [hunk 6]
```diff
@@ -130,11 +136,12 @@ services:
       EVENTUATELOCAL_CDC_MYSQL_BINLOG_CLIENT_UNIQUE_ID: 1234567890
       EVENTUATELOCAL_CDC_READ_OLD_DEBEZIUM_DB_OFFSET_STORAGE_TOPIC: "false"
       EVENTUATE_CDC_KAFKA_ENABLE_BATCH_PROCESSING: ${EVENTUATE_CDC_KAFKA_ENABLE_BATCH_PROCESSING}
+      EVENTUATE_OUTBOX_ID: 1
```
Comment (cer) [inline docker-compose-mysql-binlog.yml:139]: This should not be hardwired. Before migration it should not be defined. It should be passed using env file.

File: _build-and-test-all.sh (hunk 5 above)
Comment (dartartem) [reply _build-and-test-all.sh:40]: I will fix

File: docker-compose-mysql-binlog.yml (hunk 6 above)
Comment (dartartem) [reply docker-compose-mysql-binlog.yml:139]: I can pass it via file, but in current implementation it should be defined to work.

Comment (cer) [reply docker-compose-mysql-binlog.yml:139]: The purpose of this script is to emulate a user migrating to DB generated IDs.
Before migration, the CDC shouldn't be configured with  EVENTUATE_OUTBOX_ID
After migration, the CDC should have EVENTUATE_OUTBOX_ID=1
Isn't this the same way the services are reconfigured? Why would the CDC be different?

Comment (dartartem) [reply docker-compose-mysql-binlog.yml:139]: Chris, we had [discussion](https://github.com/eventuate-foundation/eventuate-cdc/pull/75#discussion_r494833852) about migrating cdc, the last comment is my thoughts about it (why I made EVENTUATE_OUTBOX_ID mandatory). You did not answer and later merged PR, I thought that you considered that is fine, sorry. Should I change cdc migration approach?

File: add-database-id-support-to-eventuate-mysql.migration.sql [hunk 7]
```diff
@@ -0,0 +1,43 @@
+USE eventuate;
+
+CREATE TABLE new_message (
```
Comment (cer) [inline add-database-id-support-to-eventuate-mysql.migration.sql:3]: This is a lot of copy/paste/duplication.
It should be eventuate-common/mysql

File: add-database-id-support-to-eventuate-mysql.migration.sql [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  id VARCHAR(767),
+  dbid BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,
```
Comment (cer) [inline add-database-id-support-to-eventuate-mysql.migration.sql:5]: I think the initial value of dbid should be currentTimeMillis, right?

**Comment by cer** [comment]
Merged into wip-db-id-gen branch
